src/main/java/com/conveyal/analysis/controllers/GtfsVectorTileMaker.java

```diff
@@ -0,0 +1,293 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This class maintains a spatial index of data inside GTFS feeds and produces vector tiles of it for use as a
+ * preview display. It is used by GtfsController, which defines the associated HTTP endpoints.
+ *
+ * For testing, find tile numbers with https://www.maptiler.com/google-maps-coordinates-tile-bounds-projection/
+ * Examine and analyze individual tiles with https://observablehq.com/@henrythasler/mapbox-vector-tile-dissector
+ *
+ * The tile format specification is at https://github.com/mapbox/vector-tile-spec/tree/master/2.1
+ * To summarize:
+ * Extension should be .mvt and MIME content type application/vnd.mapbox-vector-tile
+ * A Vector Tile represents data based on a square extent within a projection.
+ * Vector Tiles may be used to represent data with any projection and tile extent scheme.
+ * The reference projection is Web Mercator; the reference tile scheme is Google's.
+ * The tile should not contain information about its bounds and projection.
+ * The decoder already knows the bounds and projection of a Vector Tile it requested when decoding it.
+ * A Vector Tile should contain at least one layer, and a layer should contain at least one feature.
+ * Feature coordinates are integers relative to the top left of the tile.
+ * The extent of a tile is the number of integer units across its width and height, typically 4096.
+ *
+ * TODO handle cancellation of HTTP requests (Mapbox client cancels requests when zooming/panning)
+ */
+public class GtfsVectorTileMaker {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GtfsVectorTileMaker.class);
+
+    private final GTFSCache gtfsCache;
+
+    /**
+     * Complex street geometries will be simplified, but the geometry will not deviate from the original by more
+     * than this many tile units. These are minuscule at 1/4096 of the tile width or height.
+     */
+    private static final int LINE_SIMPLIFY_TOLERANCE = 5;
+
+    /** How long after it was last accessed to keep a GTFS spatial index in memory. */
+    private static final Duration EXPIRE_AFTER_ACCESS = Duration.ofMinutes(10);
+
+    /** The maximum number of feeds for which we keep GTFS spatial indexes in memory at once. */
+    private static final int MAX_SPATIAL_INDEXES = 4;
+
+    /** A cache of spatial indexes of the trip pattern shapes, keyed on the BundleScopedFeedId. */
+    private final LoadingCache<String, STRtree> shapeIndexCache = Caffeine.newBuilder()
+            .maximumSize(MAX_SPATIAL_INDEXES)
+            .expireAfterAccess(EXPIRE_AFTER_ACCESS)
+            .removalListener(this::logCacheEviction)
+            .build(this::buildShapesIndex);
+
+    /** A cache of spatial indexes of the transit stops, keyed on the BundleScopedFeedId. */
+    private final LoadingCache<String, STRtree> stopIndexCache = Caffeine.newBuilder()
```

Comment:
Should the shape and stop indexes be moved into the GTFSCache? If they shouldn't, should they at least be extracted into their own file in case they ever want to be used outside of the GtfsVectorTileMaker? Additionally, is there a better way to distinguish between the type of cache data they contain vs what is currently in the GTFSCache? Possibly renaming `GTFSCache` -> `GtfsMapDbCache` and extracting the shape and stop indexes into a `GtfsGeoSpatialInMemoryCache`? I don't think these names distinguish between the two well enough, but I wanted to give an example.

Reply:
Summarizing discussion on the call today: we will move these indexes into GTFSCache, but leave these spatial index caches as two top-level LoadingCaches within GTFSCache rather than making the spatial indexes fields on the GTFSFeed values managed by the GTFSCache. This is not because we necessarily consider this an ideal design long term, but because 1. it retains the current level of flexibility with respect to eviction policies, allowing the geometry caches to hold less objects and evict them sooner than the feed cache. This is important because we haven't used this in production yet and unlike the GTFSFeeds the spatial indexes have potentially unlimited memory usage. We want to observe memory usage in practice and be able to tune these parameters. 2. it's easier to reason about reference chains and loops when they're not nested, 3. even though lazy loading synchronization would be relatively simple as it would be per-GTFSFeed instance, it's nice to uniformly use the well-audited per-key locking of LoadingCache.